### DISCUSSION TITLE
POST/api/auth/login

### DISCUSSION THREAD

**Discussion by ChristiaanEmpowerBI** [MicroStrategy/rest-api-docs#39] [Q&A] [answered]
I run Windows 11, and MicroStrategy 2021 Update 11.

Can someone please tell me what is wrong with this simple code:
import mstrio
from mstrio.connection import get_connection, Connection
from mstrio.utils.error_handlers import ErrorHandler
import re
import requests
import json

PROJECT_NAME = 'Tutorial Project'  
BASE_URL = 'http://my_pc:8080/MicroStrategyLibrary/api'
MSTR_USERNAME = 'Administrator' 
MSTR_PASSWORD = ''
project_id = 'B19DEDCC11D4E0EFC000EB9495D0F44F'
dossier_id = 'FFC020B64CE9F3416D9705A368890096'

conn = Connection(
    BASE_URL,
    MSTR_USERNAME,
    MSTR_PASSWORD,
    project_id=project_id,
    login_mode=1
)

def POST_api_auth_login(connection):
    return connection.post(
        skip_expiration_check=True,
        endpoint='/api/auth/login',
        data={
            'username': connection.username,
            'password': connection._Connection__password,
            'loginMode': connection.login_mode
        },
    )

HTTP_response = POST_api_auth_login(conn) #<class 'requests.models.Response'>

conn.close()

I get error:
Connection to MicroStrategy Intelligence Server has been established.
Traceback (most recent call last):
  File "C:\Users\user\OneDrive\Documents\Python\mstrio\microstrategy_work_01.py", line 36, in <module>
    HTTP_response = POST_api_auth_login(conn) #<class 'requests.models.Response'>
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "C:\Users\user\OneDrive\Documents\Python\mstrio\microstrategy_work_01.py", line 26, in POST_api_auth_login
    return connection.post(
           ^^^^^^^^^^^^^^^^
TypeError: Connection.post() missing 1 required positional argument: 'url'

I copied and pasted this code exactly from the code snippets, and this code works on another machine. So I am stumped.

**Comment by x-du** [comment, marked as answer]
@ChristiaanEmpowerBI The parameter name is `url` not `endpoint`

**Reply by ChristiaanEmpowerBI** [reply]
Thank you for the response.

I run this code at home:
#pip install mstrio.py
#https://github.com/MicroStrategy/mstrio-py
import mstrio
from mstrio.connection import get_connection, Connection
from mstrio.utils.error_handlers import ErrorHandler
import re

PROJECT_NAME = 'Tutorial Project'  
BASE_URL = 'http://my_pc:8080/MicroStrategyLibrary/api'
MSTR_USERNAME = 'Administrator' 
MSTR_PASSWORD = ''
project_id = 'B19DEDCC11D4E0EFC000EB9495D0F44F'
dossier_id = 'C3D32F6C48EC04BACC94D281C4ABC936'

conn = Connection(
    BASE_URL,
    MSTR_USERNAME,
    MSTR_PASSWORD,
    project_id=project_id,
    login_mode=1
)

def login(connection):
    return connection.post(
        skip_expiration_check=True,
        endpoint='/api/auth/login',
        data={
            'username': connection.username,
            'password': connection._Connection__password,
            'loginMode': connection.login_mode
        },
    )

def get_JSESSIONID(input_string):
    pattern = r'JSESSIONID=([^;]+)'
    match = re.search(pattern, input_string)
    if match:
        JSESSIONID = match.group(1)
        return JSESSIONID
    else:
        print('No JSESSIONID found.')


####################### Step 1, POST /api/auth/apiTokens           ##########################
####################### connect to the Intelligence Server         ########################## 
HTTP_response = login(conn) #<class 'requests.models.Response'>
JSESSIONID = get_JSESSIONID(HTTP_response.headers['Set-Cookie'])
print('JSESSIONID:',JSESSIONID)
X_MSTR_AuthToken = HTTP_response.headers['X-MSTR-AuthToken']
print('X-MSTR-AuthToken:', X_MSTR_AuthToken)
conn.close()

and get this result:
![image](https://github.com/MicroStrategy/rest-api-docs/assets/139224196/16c35dbf-c4ab-420b-931b-69b756a1d459)

No mention of url other than BASE_URL, but it works.

**Comment by ChristiaanEmpowerBI** [comment]
I use this code at work and it works. I have no idea why at home it does not want the full url but at work it does:
#pip install mstrio.py
#https://github.com/MicroStrategy/mstrio-py
import mstrio
from mstrio.connection import get_connection, Connection
from mstrio.utils.error_handlers import ErrorHandler
import requests

PROJECT_NAME = 'Tutorial Project'  
BASE_URL = 'http://my_work_pc:8080/MicroStrategyLibrary/api'
MSTR_USERNAME = 'Administrator' 
MSTR_PASSWORD = ''
project_id = 'B19DEDCC11D4E0EFC000EB9495D0F44F'

conn = Connection(
    BASE_URL,
    MSTR_USERNAME,
    MSTR_PASSWORD,
    project_id=project_id,
    login_mode=1
)

def POST_api_auth_login(connection):
    return connection.post(
        skip_expiration_check=True,
        url='http://my_work_pc:8080/MicroStrategyLibrary/api/auth/login',
        data={
            'username': connection.username,
            'password': connection._Connection__password,
            'loginMode': connection.login_mode
        }
    )

HTTP_response = POST_api_auth_login(conn) 

conn.close()
